File: backend-go/internal/authz/roles.go
package authz

const (
	RoleStudent             = "student"
	RoleInstructor          = "instructor"
	RoleInstitutionAdmin    = "institution_admin"
	RoleContentManager      = "content_manager"
	RoleSubscriptionManager = "subscription_manager"
	RoleSuperAdmin          = "super_admin"
)

var roleHierarchy = map[string][]string{
	RoleStudent:             {},
	RoleInstructor:          {RoleStudent},
	RoleInstitutionAdmin:    {RoleStudent},
	RoleContentManager:      {RoleInstructor, RoleStudent},
	RoleSubscriptionManager: {RoleInstitutionAdmin, RoleStudent},
	RoleSuperAdmin: {
		RoleSubscriptionManager,
		RoleContentManager,
		RoleInstitutionAdmin,
		RoleInstructor,
		RoleStudent,
	},
}

func SystemRoles() []string {
	return []string{
		RoleStudent,
		RoleInstructor,
		RoleInstitutionAdmin,
		RoleContentManager,
		RoleSubscriptionManager,
		RoleSuperAdmin,
	}
}

func SelfRegisterableRoles() []string {
	return []string{
		RoleStudent,
		RoleInstructor,
	}
}

func NormalizeRoleClaims(value interface{}) []string {
	switch v := value.(type) {
	case []string:
		return v
	case []interface{}:
		roles := make([]string, 0, len(v))
		for _, item := range v {
			if role, ok := item.(string); ok {
				roles = append(roles, role)
			}
		}
		return roles
	default:
		return nil
	}
}

func ExpandRoles(roles []string) map[string]struct{} {
	expanded := make(map[string]struct{})
	var visit func(string)
	visit = func(role string) {
		if role == "" {
			return
		}
		if _, seen := expanded[role]; seen {
			return
		}
		expanded[role] = struct{}{}
		for _, inherited := range roleHierarchy[role] {
			visit(inherited)
		}
	}

	for _, role := range roles {
		visit(role)
	}

	return expanded
}

func HasAnyRole(roles []string, allowed ...string) bool {
	expanded := ExpandRoles(roles)
	for _, role := range allowed {
		if _, ok := expanded[role]; ok {
			return true
		}
	}
	return false
}

File: frontend/lib/roles.ts
export const ROLE_STUDENT = "student";
export const ROLE_INSTRUCTOR = "instructor";
export const ROLE_INSTITUTION_ADMIN = "institution_admin";
export const ROLE_CONTENT_MANAGER = "content_manager";
export const ROLE_SUBSCRIPTION_MANAGER = "subscription_manager";
export const ROLE_SUPER_ADMIN = "super_admin";

const ROLE_HIERARCHY: Record<string, string[]> = {
  [ROLE_STUDENT]: [],
  [ROLE_INSTRUCTOR]: [ROLE_STUDENT],
  [ROLE_INSTITUTION_ADMIN]: [ROLE_STUDENT],
  [ROLE_CONTENT_MANAGER]: [ROLE_INSTRUCTOR, ROLE_STUDENT],
  [ROLE_SUBSCRIPTION_MANAGER]: [ROLE_INSTITUTION_ADMIN, ROLE_STUDENT],
  [ROLE_SUPER_ADMIN]: [
    ROLE_SUBSCRIPTION_MANAGER,
    ROLE_CONTENT_MANAGER,
    ROLE_INSTITUTION_ADMIN,
    ROLE_INSTRUCTOR,
    ROLE_STUDENT,
  ],
};

export const ADMIN_ROLE_OPTIONS = [
  ROLE_STUDENT,
  ROLE_INSTRUCTOR,
  ROLE_INSTITUTION_ADMIN,
  ROLE_CONTENT_MANAGER,
  ROLE_SUBSCRIPTION_MANAGER,
  ROLE_SUPER_ADMIN,
];

export function roleLabel(role: string) {
  return role.replace(/_/g, " ");
}

export function expandRoles(roles: string[]) {
  const expanded = new Set<string>();

  const visit = (role: string) => {
    if (!role || expanded.has(role)) return;
    expanded.add(role);
    (ROLE_HIERARCHY[role] || []).forEach(visit);
  };

  roles.forEach(visit);
  return expanded;
}

export function hasAnyRole(roles: string[], allowed: string[]) {
  const expanded = expandRoles(roles);
  return allowed.some((role) => expanded.has(role));
}
